### PR TITLE
add progress callback, supress pretty_progress

### PR DESCRIPTION
**Pull request by fszontagh**

### CODE REVIEW

File: clip.hpp [hunk 1]
```diff
@@ -694,8 +694,12 @@ struct CLIPTextModel {
             LOG_ERROR("embedding '%s' failed", embd_name.c_str());
             return false;
         }
+        if (std::find(readed_embeddings.begin(), readed_embeddings.end(), embd_name) != readed_embeddings.end()) {
+            LOG_DEBUG("embedding already readed in: %s", embd_name.c_str());
```
Comment (Green-Sky) [inline clip.hpp:698]: readed -> read

File: lora.hpp [hunk 2]
```diff
@@ -54,7 +63,7 @@ struct LoraModel : public GGMLModule {
         };
 
         model_loader.load_tensors(on_new_tensor_cb, backend);
-
+
```
Comment (Green-Sky) [inline lora.hpp:66]: remove the whitespace

**Comment by leejet** [comment]
Is there a specific reason to increase the size of many graphs and some reserved buffer sizes by 2 to 4 times? Additionally, `ggml_n_dims_t` seems to serve no purpose, as `ggml_n_dims` performs the same function. Passing `n_dim` directly into the creation of `ggml_tensor` from `tensors_storage` does not seem to cause any issues, as the calculation of `ggml_tensor` dimensions does not directly use the passed `n_dim` but instead calculates it through `ggml_n_dims`.

**Comment by fszontagh** [comment]
> Is there a specific reason to increase the size of many graphs and some reserved buffer sizes by 2 to 4 times? Additionally, `ggml_n_dims_t` seems to serve no purpose, as `ggml_n_dims` performs the same function. Passing `n_dim` directly into the creation of `ggml_tensor` from `tensors_storage` does not seem to cause any issues, as the calculation of `ggml_tensor` dimensions does not directly use the passed `n_dim` but instead calculates it through `ggml_n_dims`.

There is it: #178 

But not just with the controlnet models. These errors is happened with larger file sized lora models too. I went through over my "lora collection", and tested some lora models. I modified these params while larger loras started to working. 




> `ggml_n_dims_t` seems to serve no purpose, as `ggml_n_dims`

Sorry, i don't remember what was the problem there. But the new method is required, because the original ggml_n_dims paramter type was not compatible with the type TensorStorage. 

If you want, i will try to reproducate the original problem. If i remember correctly, the tensor_storage.n_dims was empty or corrupt. 
Maybe it is no longer needed.

**Comment by fszontagh** [comment]
I reproduced the n_dims problem. If i added a lora into the prompt ([hair length slider lora](https://civitai.com/models/114215/hair-length-slider-lora)) the following assesrtion happened:
image

**Comment by leejet** [comment]
> I reproduced the n_dims problem. If i added a lora into the prompt ([hair length slider lora](https://civitai.com/models/114215/hair-length-slider-lora)) the following assesrtion happened:

@fszontagh 

I used the latest code from the master branch and did not encounter this issue. Can you try using the latest code from the master branch and see if the problem persists?

```
[INFO ] model.cpp:676  - load ..\models\hair_length_slider_v1.safetensors using safetensors format
[DEBUG] model.cpp:742  - init from '..\models\hair_length_slider_v1.safetensors'
[INFO ] lora.hpp:37   - loading LoRA from '..\models\hair_length_slider_v1.safetensors'
[DEBUG] model.cpp:1307 - loading tensors from ..\models\hair_length_slider_v1.safetensors
[DEBUG] ggml_extend.hpp:820  - lora params backend buffer size =   6.49 MB (10240 tensors)
[DEBUG] model.cpp:1307 - loading tensors from ..\models\hair_length_slider_v1.safetensors
[DEBUG] lora.hpp:67   - finished loaded lora
[DEBUG] ggml_extend.hpp:774  - lora compute buffer size: 100.02 MB
[INFO ] stable-diffusion.cpp:417  - lora 'hair_length_slider_v1' applied, taking 0.11s
```

**Comment by Cyberhan123** [comment]
I have a guess about his question. In this file: https://github.com/leejet/stable-diffusion.cpp/blob/4a8190405ac32930678ce030dff6289ed680b6fc/.gitmodules#L3C44-L3C45
It points to the official ggml git, but now the merger of this PR: https://github.com/leejet/stable-diffusion.cpp/pull/159, it actually points to leejet/ggml, this is actually confusing for most people.

**Comment by fszontagh** [comment]
It's new, because just compiled and tried to run the sd(.exe) without success. Here is the full command history
[stable_diffusion_latest_full_cmd.txt](https://github.com/leejet/stable-diffusion.cpp/files/14396944/stable_diffusion_latest_full_cmd.txt)

And here is again, but with a really full fresh start:
[stable_diffusion_latest_full_cmd_2try.txt](https://github.com/leejet/stable-diffusion.cpp/files/14396977/stable_diffusion_latest_full_cmd_2try.txt)

In wsl it's working fine with lora too. Tested with a 256,6MB lora :)

**Comment by fszontagh** [comment]
Build with MSVC 2019 with vscode, and started the diffusion, but got:
`Assertion failed: n_dims >= 1 && n_dims <= GGML_MAX_DIMS, file Z:\stable-diffusion.cpp_latest2\ggml\src\ggml.c, line 2678`

Please se the full command
[msvc2019.txt](https://github.com/leejet/stable-diffusion.cpp/files/14397116/msvc2019.txt)

**Comment by fszontagh** [comment]
Same happening here with the auto builded [release](https://github.com/Cyberhan123/stable-diffusion.cpp/releases/tag/master-346416b).
I downloaded the cuda version and tried to run, but it just stopped with a lora in the prompt.

@Cyberhan123

**Comment by fszontagh** [comment]
Another test, with the latest. I remade my changes with lora (n_dims). Then tried to reproducate an image with embedding:

`[DEBUG] model.cpp:1307 - loading tensors from D:\SD_MODELS\embeddings\ng_deepnegative_v1_75t.pt
ggml_new_object: not enough space in the context's memory pool (needed 115632, available 32768)
Assertion failed: false, file Z:\stable-diffusion.cpp_latest2\ggml\src\ggml.c, line 2643`

**Comment by leejet** [comment]
> Another test, with the latest. I remade my changes with lora (n_dims). Then tried to reproducate an image with embedding

Currently, support for very large embeddings is not available. I will add it later.

**Comment by leejet** [comment]
> Build with MSVC 2019 in vscode, and started the diffusion, but got: `Assertion failed: n_dims >= 1 && n_dims <= GGML_MAX_DIMS, file Z:\stable-diffusion.cpp_latest2\ggml\src\ggml.c, line 2678`
> 
> Please see the full command [msvc2019.txt](https://github.com/leejet/stable-diffusion.cpp/files/14397116/msvc2019.txt)
> 
> without lora it's fine

This issue is quite puzzling; I cannot replicate it in my local environment.

**Comment by fszontagh** [comment]
I build on a "virgin" PC (avx512), the same happening. Then downloaded the prebuild binary the same PC (avx512), that's working fine. But the downloaded cuda version is failed too on my machine. 

Maybe the compiler causing this? In the CI that's a enterprise version, but i use community version.

**Comment by fszontagh** [comment]
@leejet i givin up on this n_dims issue thing. But the 'progress callback' feature is a good feature. Do you accept the pr as-is? 
On summary, the ggml_n_dims_t method causing nothing bad and maybe on later it will not be necessary after some new update, but currently it's working here at least.

**Comment by Cyberhan123** [comment]
> Same happening here with the auto builded [release](https://github.com/Cyberhan123/stable-diffusion.cpp/releases/tag/master-346416b). I downloaded the cuda version and tried to run, but it just stopped with a lora in the prompt.
> 
> @Cyberhan123

I feel like your problem seems like you didn't pull the correct git submodel.

**Comment by fszontagh** [comment]
@Cyberhan123 i tested with a fresh start too

**Comment by leejet** [comment]
> But the 'progress callback' feature is a good feature. Do you accept the pr as-is?

Certainly, I'm willing to merge this PR, or we can merge the progress callback first if you prefer.

**Comment by fszontagh** [comment]
@leejet 
> Certainly, I'm willing to merge this PR, or we can merge the progress callback first if you prefer.

Okay. I pushed in it with removed ggml_n_dims_t method. After all, the 'progress callback' was the main point of this PR.

**Comment by leejet** [comment]
Thank you for your contribution.

**Pull request merged**